flink-ml-core/src/main/java/org/apache/flink/ml/builder/GraphExecutionHelper.java

```diff
@@ -0,0 +1,118 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.builder;
+
+import org.apache.flink.table.api.Table;
+import org.apache.flink.util.Preconditions;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A container class that maintains the execution state of the graph (e.g. which nodes are ready to
+ * run).
+ */
+class GraphExecutionHelper {
+    // A map from tableId to the list of nodes which take this table as input.
+    private final Map<TableId, List<GraphNode>> consumerNodes = new HashMap<>();
+    // A map from tableId to the corresponding table. The table value is null if it has not
+    // been constructed yet.
+    private final Map<TableId, Table> constructedTables = new HashMap<>();
+    // A map that maintains the number of input tables that have not been constructed
+    // for each node in the graph.
+    private final Map<GraphNode, Integer> numUnConstructedInputTables = new HashMap<>();
+    // An ordered list of nodes whose input tables have all been constructed AND who has not
+    // been fetch via pollNextReadyNode.
+    private final Deque<GraphNode> unFetchedReadyNodes = new LinkedList<>();
+
+    public GraphExecutionHelper(List<GraphNode> nodes) {
+        // Initializes dependentNodes and numUnConstructedInputs.
+        for (GraphNode node : nodes) {
+            List<TableId> inputs = new ArrayList<>();
+            inputs.addAll(Arrays.asList(node.algoOpInputIds));
+            if (node.estimatorInputIds != null) {
+                inputs.addAll(Arrays.asList(node.estimatorInputIds));
+            }
+            if (node.inputModelDataIds != null) {
+                inputs.addAll(Arrays.asList(node.inputModelDataIds));
+            }
+            numUnConstructedInputTables.put(node, inputs.size());
+            for (TableId tableId : inputs) {
+                consumerNodes.putIfAbsent(tableId, new ArrayList<>());
+                consumerNodes.get(tableId).add(node);
+            }
+        }
+    }
+
+    public void setTables(TableId[] tableIds, Table[] tables) {
+        Preconditions.checkArgument(
+                tableIds.length >= tables.length,
```

Comment:
When would `tableIds.length > tables.length` ?

Reply:
The length of `tableIds` could be higher than the length of `tables` because we over-allocate `TableIds` as placeholders of stage's outputs when building the Graph.

For example, say we have constructed a Graph where algoOperatorA's outputs are used as algoOperatorB's input. 

When `Graph::fit(...)` is invoked, what happens is that we will run algoOperatorA::fit(...) to get an array of Tables, and updates `GraphExecutionHelper` to map algoOperatorA's output TableIds to these Tables, by calling `executionHelper.setTables(node.outputIds, nodeOutputs)`. In this case, the length of `node.outputIds` would be 20 by default, by the length of `nodeOutputs` is usually less than 20.

I have added the following comment above this code to clarify it. Does this address the concern?

```
// The length of tableIds could be larger than the length of tables because we over-allocate
// the number of tableIds (which is 20 by default) as placeholder of the stage's output
// tables when adding a stage in GraphBuilder.
```